src/main/java/me/oyurimatheus/nossomercadolivre/products/ProductController.java

```diff
@@ -0,0 +1,63 @@
+package me.oyurimatheus.nossomercadolivre.products;
+
+import me.oyurimatheus.nossomercadolivre.categories.CategoryRepository;
+import me.oyurimatheus.nossomercadolivre.configuration.security.TokenManager;
+import me.oyurimatheus.nossomercadolivre.shared.ObjectIsRegisteredValidator;
+import me.oyurimatheus.nossomercadolivre.users.User;
+import me.oyurimatheus.nossomercadolivre.users.UserRepository;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.WebDataBinder;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.net.URI;
+
+import static org.springframework.http.ResponseEntity.created;
+
+@RestController
+@RequestMapping("/api/products")
+class ProductController {
+
+    private final ProductRepository productRepository;
+    private final UserRepository userRepository;
+    private final CategoryRepository categoryRepository;
+    private final PhotoUploader photoUploader;
+    private final TokenManager tokenManager;
+
+    public ProductController(ProductRepository productRepository,
+                             UserRepository userRepository,
+                             CategoryRepository categoryRepository,
+                             PhotoUploader photoUploader,
+                             TokenManager tokenManager) {
+        this.productRepository = productRepository;
+        this.userRepository = userRepository;
+        this.categoryRepository = categoryRepository;
+        this.photoUploader = photoUploader;
+        this.tokenManager = tokenManager;
+    }
+
+    @PostMapping
+    ResponseEntity<?> create(@RequestHeader(name = "Authorization") String header,
```

Comment:
*Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.*

Pq não receber um @AuthenticationPrincipal aqui já com um usuário montado?

Reply:
Hmm boa! Te confesso que não sei muitas coisas sobre o Spring security